lib/connextclient/ConnextClient.ts

```diff
@@ -0,0 +1,699 @@
+    // complete the observable and clean up all the listeners
+    // once we receive 1 event that matches our conditions
+    take(1),
+    // emit an error if the observable does not emit any values
+    // for the specified duration
+    timeout(errorTimeout),
+  );
+  // convert the observable to a promise that resolves on complete
+  // and rejects on error
+  return expectedPreimageReceived$.toPromise();
+};
+
+/**
+ * A class representing a client to interact with connext.
+ */
+class ConnextClient extends SwapClient {
+  public readonly type = SwapClientType.Connext;
+  public readonly finalLock = 200;
+  public address?: string;
+  /** A map of currency symbols to token addresses. */
+  public tokenAddresses = new Map<string, string>();
+  private expectedIncomingTransfers = new Map<string, ExpectedIncomingTransfer>();
+  private port: number;
+  private host: string;
+  private webhookport: number;
+  private webhookhost: string;
+  private unitConverter: UnitConverter;
+  private seed: string | undefined;
+
+  /**
+   * Creates a connext client.
+   */
+  constructor({
+    config,
+    logger,
+    unitConverter,
+    currencyInstances,
+  }: {
+    unitConverter: UnitConverter;
+    config: ConnextClientConfig;
+    currencyInstances: CurrencyInstance[],
+    logger: Logger;
+  }) {
+    super(logger, config.disable);
+
+    this.port = config.port;
+    this.host = config.host;
+    this.webhookhost = config.webhookhost;
+    this.webhookport = config.webhookport;
+    this.unitConverter = unitConverter;
+    this.setTokenAddresses(currencyInstances);
+  }
+
+  public get minutesPerBlock() {
+    return 0.25; // 15 seconds per block target
+  }
+
+  public get label() {
+    return 'Connext';
+  }
+
+  public initSpecific = async () => {
+    // listen for incoming transfers
+    this.on('transferReceived', (transferReceivedRequest: TransferReceivedEvent) => {
+      const {
+        tokenAddress,
+        units,
+        timelock,
+        rHash,
+      } = transferReceivedRequest;
+      const expectedIncomingTransfer = this.expectedIncomingTransfers.get(rHash);
+      if (!expectedIncomingTransfer) {
+        this.logger.error(`received unexpected incoming transfer with rHash ${rHash}`);
+        return;
+      }
+      const {
+        rHash: expectedHash,
+        units: expectedUnits,
+        expiry: expectedTimelock,
+        tokenAddress: expectedTokenAddress,
+      } = expectedIncomingTransfer;
+      const currency = this.getCurrencyByTokenaddress(tokenAddress);
+      if (
+        units === expectedUnits &&
+        rHash === expectedHash &&
+        timelock === expectedTimelock &&
+        tokenAddress === expectedTokenAddress
+      ) {
+        this.logger.debug(`accepting incoming transfer with rHash: ${rHash}, units: ${units}, timelock ${timelock} and currency ${currency}`);
+        this.emit('htlcAccepted', rHash, units, currency);
+        // TODO: should we also store this in the database?
```

Comment:
I also have this question. Should we even be storing all expected incoming transfers into the database? If xud crashes while we are expecting transfers, would we need them when we come back online?

Reply:
> Should we even be storing all expected incoming transfers into the database?

I think we can derive the expected incoming transfers from the `SwapDeal` since we're persisting every step of the deal.

> If xud crashes while we are expecting transfers, would we need them when we come back online?

Definitely.

Reply:
After thinking about it for a while... should the SwapRecovery flow handle that instead?